pkg/network/payload/address.go

```diff
@@ -30,26 +29,30 @@ func NewAddressAndTime(e *net.TCPAddr, t time.Time) *AddressAndTime {
 // DecodeBinary implements Serializable interface.
 func (p *AddressAndTime) DecodeBinary(br *io.BinReader) {
 	p.Timestamp = br.ReadU32LE()
-	p.Services = br.ReadU64LE()
 	br.ReadBytes(p.IP[:])
-	p.Port = br.ReadU16BE()
+	p.Capabilities.DecodeBinary(br)
 }
 
 // EncodeBinary implements Serializable interface.
 func (p *AddressAndTime) EncodeBinary(bw *io.BinWriter) {
 	bw.WriteU32LE(p.Timestamp)
-	bw.WriteU64LE(p.Services)
 	bw.WriteBytes(p.IP[:])
-	bw.WriteU16BE(p.Port)
+	p.Capabilities.EncodeBinary(bw)
 }
 
 // IPPortString makes a string from IP and port specified.
 func (p *AddressAndTime) IPPortString() string {
 	var netip = make(net.IP, 16)
 
 	copy(netip, p.IP[:])
-	port := strconv.Itoa(int(p.Port))
-	return netip.String() + ":" + port
+	var tcpPort uint16
+	for _, cap := range p.Capabilities {
+		if cap.Type == capability.TCPServer {
+			tcpPort = cap.Data.(*capability.Server).Port
+			break
+		}
+	}
+	return net.JoinHostPort(netip.String(), strconv.Itoa(int(tcpPort)))
```

Comment:
So it's gonna be "a.b.c.d:0" when no `TCPServer` capability was sent? Not sure that's what we want.

Reply:
Added also port from `WSServer`. `TCPServer` and `WSServer` are the only capabilities we're able to get port from. What's the other way to do this?

Reply:
> Added also port from WSServer

NACK.

The answer to this is in the way this `IPPortString()` is being used and it's being used to add some hosts for discovery service. We're operating as `TCPServer` and we only want to make friends with other `TCPServers`, so we should only add there hosts that have proper `TCPServer` capability set. Pushing `WSServer` port there is just plain wrong, we won't be able to connect to that.

So in the end we could transform it into `GetTCPAddress() (string, error)` and return error when this `AddressAndTime` doesn't have the respective capability. And this error could be ignored in `handleAddrCmd` because it's OK for a host to not be a `TCPServer`.

Reply:
Done.